### PR TITLE
FOGL-8771: Fixed undo/redo functionality for flow editor canvas

### PR DESCRIPTION
**Pull request by hemantgarg923**

### CODE REVIEW

File: src/app/components/common/node-editor/editor.ts [hunk 1]
```diff
@@ -41,6 +41,7 @@ class Connection<A extends Node, B extends Node> extends ClassicPreset.Connectio
 
 let editor = new NodeEditor<Schemes>();
 let area: AreaPlugin<Schemes, AreaExtra>;
+let history = new HistoryPlugin<Schemes, HistoryActions<Schemes>>({timing: 100});
```
Comment (praveen-garg) [inline src/app/components/common/node-editor/editor.ts:44]: What is the meaning of 100 here?

**Review by praveen-garg — DISMISSED** [review]

File: src/app/components/common/node-editor/editor.ts (hunk 1 above)
Comment (hemantgarg923) [reply src/app/components/common/node-editor/editor.ts:44]: This 100ms is actions grouping time. There will be a single entry in history array for all the actions occurring within this time duration. So after pressing Ctrl+z all actions occurred within this time duration will be reverted at once. Default value for this is 200ms.

**Review by MonikaSharma06 — APPROVED** [review]

File: src/app/components/common/node-editor/editor.ts [hunk 2]
```diff
@@ -570,3 +574,53 @@ export function applyContentReordering(nodeId: string) {
...
+      dock.add(newDockFilter);
+    }
+  })
+}
+
+function getSecondLastActionName() {
+  let historySnapshot = history.getHistorySnapshot();
+  let historyLength = historySnapshot.length;
+  let actionName;
+  if (historyLength >= 2) {
```
Comment (mshariq-nerd) [inline src/app/components/common/node-editor/editor.ts:610]: Why history length > 2 required?

File: src/app/components/common/node-editor/editor.ts [hunk 3]
```diff
@@ -570,3 +574,53 @@ export function applyContentReordering(nodeId: string) {
...
+    // FIXME: use different approach for retrieving actionName
+    actionName = Object.getPrototypeOf(historySnapshot[historyLength - 2].action).constructor.name;
```
Comment (mshariq-nerd) [inline src/app/components/common/node-editor/editor.ts:612]: This code block will not work in the production. Better to comment this?

File: src/app/components/common/node-editor/editor.ts [hunk 4]
```diff
@@ -570,3 +574,53 @@ export function applyContentReordering(nodeId: string) {
...
+  return actionName;
+}
+
+function getLastActionName() {
+  let historySnapshot = history.getHistorySnapshot();
+  let historyLength = historySnapshot.length;
+  let actionName;
+  if (historyLength >= 1) {
+    // FIXME: use different approach for retrieving actionName
+    actionName = Object.getPrototypeOf(historySnapshot[historyLength - 1].action).constructor.name;
```
Comment (mshariq-nerd) [inline src/app/components/common/node-editor/editor.ts:623]: As above

File: src/app/components/common/node-editor/editor.ts (hunk 2 above)
Comment (hemantgarg923) [reply src/app/components/common/node-editor/editor.ts:610]: Adding a node from dock menu creates 2 entries in history array. First is for `AddNodeAction` and second is for `DragNodeAction`. so need to check the length of array before retrieving actionName.

**Review by mshariq-nerd — APPROVED** [review]

**Pull request merged**